text/204-golang-bindings.md

```diff
@@ -1068,12 +1068,249 @@ This allows construction like so:
+the `reflect` standard library package.
+
+```go
+// Given this struct
+type WithRequiredProperties struct {
+    Required     InterfaceType
+    Optional     Option[InterfaceType]
+}
+
+// The following are all valid, compiler-wise
+a := WithRequiredProperties{}
+b := WithRequiredProperties{Optional: someValue}
+c := WithRequiredProperties{Required: someValue}
+d := WithRequiredProperties{Required: someValue, optional: someValue}
+
+// When that struct is passed over the wire to node, we can determine that a and b are invalid
+SomeStatic(a) // #=> Panic ("missing value for required property 'Required'")
+SomeStatic(b) // #=> Panic ("missing value for required property 'Required'")
+SomeStatic(c) // #=> OK
+SomeStatic(d) // #=> OK
+
+// Without `Option[T]`, this can only be achieved by generating a validator function, and calling that on EVERY usage
+// site of the struct.
+```
+
+Not using pointers everywhere also enables the runtime library to do away with a lot of pointer indirection work, as
+values may no longer be of pointer types and instead are always either a plain struct (in the case of jsii structs), or
+an interface type. This results in simplifications in some areas of the runtime library.
+
+Non-optional jsii struct parameters are passed as go structs, which provides compile-time gurantee that the value is not
+nil (but does not gurantee all fields have been set to a value other than their zero-value). Thi is more idiomatic go
+than the option of passing pointers to structs.
+
+```go
+type MyStruct struct { /* ... */ }
+
+// Using pointers:
+func WithPointers(ptr *MyStruct) { /* ... */ }
+WithPointers(nil) // Compiles, but is invalid!
+
+// Using bare struct
+func WithStruct(val MyStruct) { /* ... */ }
+withPointers(nil) // Compile error
+```
+
+An "unchecked cast" feature is much nicer to use when using generics than when those are not available:
+
+```go
+// Without generics
+func UncheckedCast(from interface{}, to *interface{}) {
+    // Must use reflection & pointer indirection to determine the target type
+    t := reflect.TypeOf(to).Elem()
+    // ... etc ...
+}
+
+opaqueValue := GetOpaque() // Typed interface{} or a parent type of the dynamic type of the value
+var castValue SpecificType // Must declare separate variable, then "cast into it"
+UncheckedCast(opaqueValue, &castValue)
+
+// With generics
+func UncheckedCast[T any](from interface{}) T {
+    var result T // Cast function can internally create the target value
+    // ...
+    return result
+}
+
+// No need to declare a separate variable up-front here.
+UncheckedCast[SpecificType](opaqueValue)
+```
+
+### Downsides
+
+In go, `nil` is a valid value at any location where an interface is expected. This means it is not possible to achieve
+compile-time safety with this approach in places where classes are passed as arguments, since those are represented as
```

Comment:
is this not the current situation as well? Where a field in TS is type `SomeClass` (required) and in go is `SomeClass`, the interface is met by null, though the property is required?

Reply:
Correct. I just wanted to be very explicit about the guarantees the compiler will NOT provide.